File: codes/jiheon/649_dota2senate.java
class Solution {
    public String predictPartyVictory(String senate) {
        int rcnt = 0, dcnt = 0, rscore = 0;
        var dq = new ArrayDeque<Character>();
        for (var c : senate.toCharArray()) {
            if (c == 'R')rcnt++;
            else dcnt++;
            dq.add(c);
        }

        while (rcnt > 0 && dcnt > 0) {
            var tmp = dq.pollFirst();
            if (tmp == 'R') {
                if (rscore < 0)rcnt--;
                else dq.addLast(tmp);
                rscore++;
            } else {
                if (rscore > 0)dcnt--;
                else dq.addLast(tmp);
                rscore--;
            }
        }
        return rcnt == 0 ? "Dire" : "Radiant";
    }
}

File: codes/jiheon/649_dota2senate.kt
class Solution {
    fun predictPartyVictory(senate: String): String {

        val charArray = senate.toCharArray()
        var rScore = 0;

        fun helper(length: Int = senate.length): String {
            var idx = 0;

            for (i in 0 until length) {
                val c = charArray[i];

                if (c == 'R') {
                    if (rScore >= 0) charArray[idx++] = 'R'
                    rScore++
                } else {
                    if (rScore <= 0) charArray[idx++] = 'D'
                    rScore--;
                }
            }

            if (abs(rScore) >= idx) return if (rScore > 0) "Radiant" else "Dire"
            
            return helper(idx)
        }
        return helper()
    }
}